[Bug][Failing Test]: Distributed Comm Ops - distributed/test_shm_broadcast.py

### Your current environment

pytest -v -x distributed/test_shm_broadcast.py

https://buildkite.com/vllm/ci/builds/20415#0196f100-f85c-4db6-8b50-72d3d5ade137/197-990


### 🐛 Describe the bug

See above

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

Test failure:

```
FAILED distributed/test_shm_broadcast.py::test_shm_broadcast - AssertionError: assert 1 == 0
```

<detail><summary>Logs:</summary>
```
Rank 2 got seed 752
Rank 0 got seed 752
Rank 1 got seed 752
StatelessProcessGroup passed the test!
StatelessProcessGroup passed the test!
StatelessProcessGroup passed the test!
[rank1]:[W520 21:12:22.801158701 TCPStore.cpp:115] [c10d] recvVector failed on SocketImpl(fd=24, addr=[ip-172-16-0-2.us-west-2.compute.internal]:32772, remote=[ip-172-16-0-2.us-west-2.compute.internal]:49587): Connection reset by peer
Exception raised from recvBytes at /pytorch/torch/csrc/distributed/c10d/Utils.hpp:675 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x98 (0x7f06545785e8 in /usr/local/lib/python3.12/dist-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x5ba8afe (0x7f063863cafe in /usr/local/lib/python3.12/dist-packages/torch/lib/libtorch_cpu.so)
frame #2: <unknown function> + 0x5baa15f (0x7f063863e15f in /usr/local/lib/python3.12/dist-packages/torch/lib/libtorch_cpu.so)
frame #3: <unknown function> + 0x5baa81d (0x7f063863e81d in /usr/local/lib/python3.12/dist-packages/torch/lib/libtorch_cpu.so)
frame #4: <unknown function> + 0x5bab4a9 (0x7f063863f4a9 in /usr/local/lib/python3.12/dist-packages/torch/lib/libtorch_cpu.so)
frame #5: c10d::TCPStore::doGet(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xbd (0x7f0638639f2d in /usr/local/lib/python3.12/dist-packages/torch/lib/libtorch_cpu.so)
frame #6: c10d::TCPStore::get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xab (0x7f063863af8b in /usr/local/lib/python3.12/dist-packages/torch/lib/libtorch_cpu.so)
frame #7: <unknown function> + 0xc0f526 (0x7f064796d526 in /usr/local/lib/python3.12/dist-packages/torch/lib/libtorch_python.so)
frame #8: <unknown function> + 0x37f17d (0x7f06470dd17d in /usr/local/lib/python3.12/dist-packages/torch/lib/libtorch_python.so)
frame #9: /usr/bin/python3() [0x56ad3d]
frame #10: _PyObject_MakeTpCall + 0x2fb (0x53d40b in /usr/bin/python3)
frame #11: _PyEval_EvalFrameDefault + 0x700 (0x548140 in /usr/bin/python3)
frame #12: _PyObject_FastCallDictTstate + 0x1d8 (0x53fca8 in /usr/bin/python3)
frame #13: /usr/bin/python3() [0x57bf2a]
frame #14: _PyObject_MakeTpCall + 0x2bb (0x53d3cb in /usr/bin/python3)
frame #15: _PyEval_EvalFrameDefault + 0x700 (0x548140 in /usr/bin/python3)
frame #16: _PyObject_FastCallDictTstate + 0x256 (0x53fd26 in /usr/bin/python3)
frame #17: _PyObject_Call_Prepend + 0x59 (0x57c3e9 in /usr/bin/python3)
frame #18: /usr/bin/python3() [0x6673bd]
frame #19: _PyObject_MakeTpCall + 0x2fb (0x53d40b in /usr/bin/python3)
frame #20: _PyEval_EvalFrameDefault + 0x700 (0x548140 in /usr/bin/python3)
frame #21: _PyObject_FastCallDictTstate + 0x256 (0x53fd26 in /usr/bin/python3)
frame #22: _PyObject_Call_Prepend + 0x59 (0x57c3e9 in /usr/bin/python3)
frame #23: /usr/bin/python3() [0x6673bd]
frame #24: _PyObject_Call + 0x93 (0x57f023 in /usr/bin/python3)
frame #25: _PyEval_EvalFrameDefault + 0x4f52 (0x54c992 in /usr/bin/python3)
frame #26: _PyObject_FastCallDictTstate + 0x256 (0x53fd26 in /usr/bin/python3)
frame #27: _PyObject_Call_Prepend + 0x59 (0x57c3e9 in /usr/bin/python3)
frame #28: /usr/bin/python3() [0x6673bd]
frame #29: _PyObject_MakeTpCall + 0x2fb (0x53d40b in /usr/bin/python3)
frame #30: _PyEval_EvalFrameDefault + 0x700 (0x548140 in /usr/bin/python3)
frame #31: _PyObject_FastCallDictTstate + 0x256 (0x53fd26 in /usr/bin/python3)
frame #32: _PyObject_Call_Prepend + 0x59 (0x57c3e9 in /usr/bin/python3)
frame #33: /usr/bin/python3() [0x6673bd]
frame #34: _PyObject_MakeTpCall + 0x2fb (0x53d40b in /usr/bin/python3)
frame #35: _PyEval_EvalFrameDefault + 0x700 (0x548140 in /usr/bin/python3)
frame #36: _PyObject_FastCallDictTstate + 0x256 (0x53fd26 in /usr/bin/python3)
frame #37: _PyObject_Call_Prepend + 0x59 (0x57c3e9 in /usr/bin/python3)
frame #38: /usr/bin/python3() [0x6673bd]
frame #39: _PyObject_MakeTpCall + 0x2fb (0x53d40b in /usr/bin/python3)
frame #40: _PyEval_EvalFrameDefault + 0x700 (0x548140 in /usr/bin/python3)
frame #41: PyEval_EvalCode + 0x99 (0x61d5b9 in /usr/bin/python3)
frame #42: /usr/bin/python3() [0x6591db]
frame #43: /usr/bin/python3() [0x654346]
frame #44: /usr/bin/python3() [0x651295]
frame #45: _PyRun_SimpleFileObject + 0x1a5 (0x650c85 in /usr/bin/python3)
frame #46: _PyRun_AnyFileObject + 0x47 (0x650a87 in /usr/bin/python3)
frame #47: Py_RunMain + 0x375 (0x64d4e5 in /usr/bin/python3)
frame #48: Py_BytesMain + 0x2d (0x6064ad in /usr/bin/python3)
frame #49: <unknown function> + 0x29d90 (0x7f06d5bc7d90 in /lib/x86_64-linux-gnu/libc.so.6)
frame #50: __libc_start_main + 0x80 (0x7f06d5bc7e40 in /lib/x86_64-linux-gnu/libc.so.6)
frame #51: _start + 0x25 (0x606325 in /usr/bin/python3)

Process Process-2:
Traceback (most recent call last):
  File "/usr/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.12/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/vllm-workspace/tests/distributed/test_shm_broadcast.py", line 52, in wrapped_fn
    fn()
  File "/vllm-workspace/tests/distributed/test_shm_broadcast.py", line 112, in worker_fn
    pg.barrier()
  File "/usr/local/lib/python3.12/dist-packages/vllm/distributed/utils.py", line 218, in barrier
    self.broadcast_obj(None, src=i)
  File "/usr/local/lib/python3.12/dist-packages/vllm/distributed/utils.py", line 199, in broadcast_obj
    recv_obj = pickle.loads(self.store.get(key))
                            ^^^^^^^^^^^^^^^^^^^
torch.distributed.DistNetworkError: Connection reset by peer
FAILED

=================================== FAILURES ===================================
______________________________ test_shm_broadcast ______________________________

    def test_shm_broadcast():
>       distributed_run(worker_fn, 4)

distributed/test_shm_broadcast.py:117: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

fn = <function worker_fn_wrapper.<locals>.wrapped_fn at 0x7f04831bbf60>
world_size = 4

    def distributed_run(fn, world_size):
        number_of_processes = world_size
        processes = []
        for i in range(number_of_processes):
            env = {}
            env['RANK'] = str(i)
            env['LOCAL_RANK'] = str(i)
            env['WORLD_SIZE'] = str(number_of_processes)
            env['LOCAL_WORLD_SIZE'] = str(number_of_processes)
            env['MASTER_ADDR'] = 'localhost'
            env['MASTER_PORT'] = '12345'
            p = multiprocessing.Process(target=fn, args=(env, ))
            processes.append(p)
            p.start()
    
        for p in processes:
            p.join()
    
        for p in processes:
>           assert p.exitcode == 0
E           AssertionError: assert 1 == 0
E            +  where 1 = <Process name='Process-2' pid=34731 parent=19 stopped exitcode=1>.exitcode

distributed/test_shm_broadcast.py:42: AssertionError
=============================== warnings summary ===============================
../../usr/local/lib/python3.12/dist-packages/schemathesis/generation/coverage.py:305
  /usr/local/lib/python3.12/dist-packages/schemathesis/generation/coverage.py:305: DeprecationWarning: jsonschema.exceptions.RefResolutionError is deprecated as of version 4.18.0. If you wish to catch potential reference resolution errors, directly catch referencing.exceptions.Unresolvable.
    ref_error: type[Exception] = jsonschema.RefResolutionError,

tests/distributed/test_shm_broadcast.py::test_shm_broadcast
tests/distributed/test_shm_broadcast.py::test_shm_broadcast
tests/distributed/test_shm_broadcast.py::test_shm_broadcast
tests/distributed/test_shm_broadcast.py::test_shm_broadcast
  /usr/lib/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=19) is multi-threaded, use of fork() may lead to deadlocks in the child.
    self.pid = os.fork()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED distributed/test_shm_broadcast.py::test_shm_broadcast - AssertionError: assert 1 == 0
 +  where 1 = <Process name='Process-2' pid=34731 parent=19 stopped exitcode=1>.exitcode
======================== 1 failed, 5 warnings in 15.52s ========================
^^^ +++
🚨 Error: The command exited with status 1
^^^ +++
user command error: The plugin docker command hook exited with status 1
~~~ Running global pre-exit hook
$ /etc/buildkite-agent/hooks/pre-exit
~~~ Running plugin docker pre-exit hook
$ /var/lib/buildkite-agent/plugins/bk-gpu-4-queue-ci-i-041889359d2c9cb88-1/github-com-buildkite-plugins-docker-buildkite-plugin-v5-2-0/hooks/pre-exit
```
</detail>

--- a/tests/distributed/test_shm_broadcast.py
+++ b/tests/distributed/test_shm_broadcast.py
@@ -9,7 +9,7 @@ import torch.distributed as dist
 
 from vllm.distributed.device_communicators.shm_broadcast import MessageQueue
 from vllm.distributed.utils import StatelessProcessGroup
-from vllm.utils import get_ip, get_open_port, update_environment_variables
+from vllm.utils import get_open_port, update_environment_variables
 
 
 def get_arrays(n: int, seed: int = 0) -> list[np.ndarray]:
@@ -60,12 +60,12 @@ def worker_fn():
     rank = dist.get_rank()
     if rank == 0:
         port = get_open_port()
-        ip = get_ip()
+        ip = '127.0.0.1'
         dist.broadcast_object_list([ip, port], src=0)
     else:
         recv = [None, None]
         dist.broadcast_object_list(recv, src=0)
-        ip, port = recv
+        ip, port = recv  # type: ignore
 
     stateless_pg = StatelessProcessGroup.create(ip, port, rank,
                                                 dist.get_world_size())
@@ -107,10 +107,10 @@ def worker_fn():
 
         if pg == dist.group.WORLD:
             dist.barrier()
-            print("torch distributed passed the test!")
+            print(f"torch distributed passed the test! Rank {rank}")
         else:
             pg.barrier()
-            print("StatelessProcessGroup passed the test!")
+            print(f"StatelessProcessGroup passed the test! Rank {rank}")
 
 
 def test_shm_broadcast():

--- a/vllm/distributed/device_communicators/shm_broadcast.py
+++ b/vllm/distributed/device_communicators/shm_broadcast.py
@@ -1,8 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0
 
-import os
 import pickle
-import sys
 import time
 from contextlib import contextmanager
 from dataclasses import dataclass, field
@@ -19,7 +17,7 @@ from zmq import IPV6  # type: ignore
 from zmq import SUB, SUBSCRIBE, XPUB, XPUB_VERBOSE, Context  # type: ignore
 
 import vllm.envs as envs
-from vllm.distributed.utils import StatelessProcessGroup
+from vllm.distributed.utils import StatelessProcessGroup, sched_yield
 from vllm.logger import init_logger
 from vllm.utils import (get_ip, get_open_port, get_open_zmq_ipc_path,
                         is_valid_ipv6_address)
@@ -27,20 +25,6 @@ from vllm.utils import (get_ip, get_open_port, get_open_zmq_ipc_path,
 VLLM_RINGBUFFER_WARNING_INTERVAL = envs.VLLM_RINGBUFFER_WARNING_INTERVAL
 
 logger = init_logger(__name__)
-
-# We prefer to use os.sched_yield as it results in tighter polling loops,
-# measured to be around 3e-7 seconds. However on earlier versions of Python
-# os.sched_yield() does not release the GIL, so we fall back to time.sleep(0)
-USE_SCHED_YIELD = ((sys.version_info[:3] >= (3, 11, 1))
-                   or (sys.version_info[:2] == (3, 10)
-                       and sys.version_info[2] >= 8))
-
-
-def sched_yield():
-    if USE_SCHED_YIELD:
-        os.sched_yield()
-    else:
-        time.sleep(0)
 
 
 class ShmRingBuffer:

--- a/vllm/distributed/utils.py
+++ b/vllm/distributed/utils.py
@@ -6,9 +6,12 @@
 # Copyright (c) 2022, NVIDIA CORPORATION. All rights reserved.
 import dataclasses
 import datetime
+import os
 import pickle
 import socket
+import sys
 import time
+import uuid
 from collections import deque
 from collections.abc import Sequence
 from typing import Any, Optional
@@ -26,6 +29,20 @@ from vllm.logger import init_logger
 from vllm.utils import get_tcp_uri, is_torch_equal_or_newer
 
 logger = init_logger(__name__)
+
+# We prefer to use os.sched_yield as it results in tighter polling loops,
+# measured to be around 3e-7 seconds. However on earlier versions of Python
+# os.sched_yield() does not release the GIL, so we fall back to time.sleep(0)
+USE_SCHED_YIELD = ((sys.version_info[:3] >= (3, 11, 1))
+                   or (sys.version_info[:2] == (3, 10)
+                       and sys.version_info[2] >= 8))
+
+
+def sched_yield():
+    if USE_SCHED_YIELD:
+        os.sched_yield()
+    else:
+        time.sleep(0)
 
 
 def ensure_divisibility(numerator, denominator):
@@ -212,10 +229,141 @@ class StatelessProcessGroup:
                 gathered_objs.append(recv_obj)
         return gathered_objs
 
-    def barrier(self):
-        """A barrier to synchronize all ranks."""
+    def barrier(self, timeout: float = 30.0):
+        """A robust barrier to synchronize all ranks.
+
+
+        Uses a multi-phase approach to ensure all processes reach the barrier
+        before proceeding:
+
+        1. Each process signals it has reached the barrier
+
+        2. Each process signals that it has confirmed the arrival of all other
+        ranks.
+
+        3. Rank 0 waits for all other ranks to signal their departure to ensure
+        that all ranks have departed the barrier first.
+
+        Args:
+            timeout: Maximum time in seconds to wait for each phase (in seconds)
+
+
+        Raises:
+            RuntimeError: If coordination fails or times out
+        """
+        # Generate a barrier ID that is globally unique
+        try:
+            if self.rank == 0:
+                barrier_id = f"barrier_{uuid.uuid4()}"
+                self.broadcast_obj(barrier_id, src=0)
+            else:
+                barrier_id = self.broadcast_obj(None, src=0)
+        except Exception as e:
+            raise RuntimeError("Failed to broadcast barrier_id") from e
+
+        # Phase 1: Signal arrival at barrier
+        # Wait for all processes to arrive
+        # We need all ranks to confirm the arrival of all other ranks.
+        # This is the key synchronization point.
+        arrival_key = f"arrival_{barrier_id}_{self.rank}"
+        try:
+            self.store.set(arrival_key, b"1")
+        except Exception as e:
+            raise RuntimeError("Failed to signal barrier arrival") from e
+
+        start_time = time.time()
+        processes_arrived: set[int] = set()
+
+        while len(processes_arrived) < self.world_size:
+            # Check for timeout
+            cur_time = time.time()
+            if cur_time - start_time > timeout:
+                raise RuntimeError("Barrier timed out after %f seconds",
+                                   timeout)
+
+            # Check for each process
+            for i in range(self.world_size):
+                if i in processes_arrived:
+                    continue
+
+                key = f"arrival_{barrier_id}_{i}"
+                try:
+                    # Try to get the key - if it exists, we'll get a value
+                    # If it doesn't exist, it will throw an exception
+                    self.store.get(key)
+                    processes_arrived.add(i)
+                except KeyError:
+                    # Key doesn't exist yet
+                    pass
+                except Exception as check_e:
+                    logger.debug("Error checking key existence: %s", check_e)
+                    sched_yield()
+
+            # Short sleep to avoid tight polling
+            if len(processes_arrived) < self.world_size:
+                sched_yield()
+
+        # Phase 2: Signal departure from barrier
+        # We only care to block at this stage in rank 0, which runs the
+        # server side of the TCPStore. We want to make sure that all
+        # clients have departed the barrier before rank 0 in case the
+        # next thing after the barrier is a shutdown, including tearing
+        # down the TCPStore. Other ranks can exit the barrier immediately
+        # after signaling their departure.
+        departure_key = f"departure_{barrier_id}_{self.rank}"
+        try:
+            self.store.set(departure_key, b"1")
+        except Exception as e:
+            raise RuntimeError("Failed to signal barrier departure") from e
+
+        if self.rank != 0:
+            return
+
+        # Make rank 0 wait for all processes to signal departure
+        start_time = time.time()
+        processes_departed: set[int] = set()
+
+        while len(processes_departed) < self.world_size:
+            # Check for timeout
+            if time.time() - start_time > timeout:
+                raise RuntimeError("Barrier departure timed out after %f s",
+                                   timeout)
+
+            # Check for each process
+            for i in range(self.world_size):
+                if i in processes_departed:
+                    continue
+
+                key = f"departure_{barrier_id}_{i}"
+                try:
+                    # Try to get the key - if it exists, we'll get a value
+                    # If it doesn't exist, it will throw an exception
+                    self.store.get(key)
+                    processes_departed.add(i)
+                except KeyError:
+                    # Key doesn't exist yet
+                    pass
+                except Exception as check_e:
+                    logger.debug("Error checking key existence: %s", check_e)
+                    sched_yield()
+
+            # Short sleep to avoid tight polling
+            if len(processes_departed) < self.world_size:
+                sched_yield()
+
+        # Clean up keys to avoid leaking memory in the store
         for i in range(self.world_size):
-            self.broadcast_obj(None, src=i)
+            try:
+                self.store.delete_key(f"arrival_{barrier_id}_{i}")
+            except Exception:
+                logger.debug("Error deleting key: %s",
+                             f'arrival_{barrier_id}_{i}')
+
+            try:
+                self.store.delete_key(f"departure_{barrier_id}_{i}")
+            except Exception:
+                logger.debug("Error deleting key: %s",
+                             f'departure_{barrier_id}_{i}')
 
     @staticmethod
     def create(

